crates/rtsan-sys/build.rs

```diff
@@ -0,0 +1,181 @@
+use std::env;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use tempfile::tempdir;
+
+fn main() {
+    #[cfg(not(any(target_os = "macos", target_os = "linux")))]
+    {
+        println!("cargo:warning=rtsan is currently only supported on macOS and Linux.");
+        return;
+    }
+
+    let target_os = env::var("CARGO_CFG_TARGET_OS").unwrap();
+    let target_arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap();
+    let out_dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+
+    // Check if pre-built library path is provided
+    if let Ok(prebuilt_lib) = env::var("RTSAN_LIB_PATH") {
+        let prebuilt_path = PathBuf::from(prebuilt_lib);
+        if !prebuilt_path.exists() {
+            panic!("Provided library path does not exist: {:?}", prebuilt_path);
+        }
+
+        let expected_extension = if target_os == "linux" { "a" } else { "dylib" };
+        if !prebuilt_path
+            .extension()
+            .map_or(false, |ext| ext == expected_extension)
+        {
+            panic!("Invalid library extension for target OS");
+        }
+
+        // Copy the library to OUT_DIR
+        let lib_name = prebuilt_path.file_name().unwrap();
+        let dest_lib_path = out_dir.join(lib_name);
+        fs::copy(&prebuilt_path, &dest_lib_path).expect("Failed to copy library to OUT_DIR");
+
+        setup_linking(&dest_lib_path, &target_os);
+        return;
+    }
+
+    // If no pre-built library provided, build from source
+    check_tool("git");
+    check_tool("cmake");
+    check_tool("make");
+
+    if target_os == "macos" {
+        check_tool("install_name_tool");
+    }
+
+    // Create a unique temporary directory
+    let temp_dir = tempdir().expect("Failed to create temporary directory");
+    let llvm_project_dir = temp_dir.path().join("llvm-project");
+
+    // Clone llvm-project into the temporary directory
+    run_command(
+        "git",
+        &[
+            "clone",
+            "-n",
+            "--depth=1",
+            "--filter=tree:0",
+            "https://github.com/llvm/llvm-project.git",
+            llvm_project_dir.to_str().unwrap(),
+        ],
+        ".",
+    );
+
+    // Perform sparse checkout
+    run_command(
+        "git",
+        &[
+            "sparse-checkout",
+            "set",
+            "--no-cone",
+            "compiler-rt",
+            "cmake",
+        ],
+        llvm_project_dir.to_str().unwrap(),
+    );
```

Comment:
Not a note for this review, but just for visibility: When the official LLVM 20 gets released we should pin that here, just so we aren't constantly consuming the main branch and getting whatever garbage may be floating in there.